cloud/cluster/cluster.go

```diff
@@ -5,121 +5,153 @@ package cluster
 func (c *Cluster) loop() {
-	for !c.done() {
-		select {
-		case nodesOrUpdates, ok := <-c.updateCh:
-			if !ok {
-				c.updateCh = nil
-				continue
-			}
-			outgoing := []NodeUpdate{}
-			if updates, ok := nodesOrUpdates.([]NodeUpdate); ok {
-				outgoing = c.state.filterAndUpdate(updates)
-			} else if nodes, ok := nodesOrUpdates.([]Node); ok {
-				sort.Sort(NodeSorter(nodes))
-				outgoing = c.state.setAndDiff(nodes)
-			}
-			for _, sub := range c.subs {
-				sub <- outgoing
+	// start processing loop
+	ticker := time.NewTicker(c.updateFreq)
+	for range ticker.C {
+		lastestNodeList := c.getLatestNodesList() // get latest fetched nodes
+		if lastestNodeList != nil {
+			sort.Sort(NodeSorter(lastestNodeList))
+			updates := c.state.setAndDiff(lastestNodeList) // compute updates and update local state
+			if c.NodesUpdatesCh != nil {
+				c.addToCurrentNodeUpdates(updates)
 			}
-		case req, ok := <-c.reqCh:
-			if !ok {
-				c.reqCh = nil
-				continue
+		}
+
+		if c.NodesReqCh != nil {
+			select {
+			case respCh := <-c.NodesReqCh:
+				respCh <- c.getNodes()
+			default:
 			}
-			c.handleReq(req)
 		}
 	}
-	for _, sub := range c.subs {
-		close(sub)
-	}
 }
 
-func (c *Cluster) handleReq(req interface{}) {
-	switch req := req.(type) {
-	case chan []Node:
-		// Members()
-		req <- c.current()
-	case chan Subscription:
-		// Subscribe()
-		ch := make(chan []NodeUpdate)
-		s := makeSubscription(c.current(), c, ch)
-		c.subs = append(c.subs, ch)
-		req <- s
-	case chan []NodeUpdate:
-		// close of a subscription
-		for i, sub := range c.subs {
-			if sub == req {
-				c.subs = append(
-					c.subs[0:i],
-					c.subs[i+1:]...)
-				close(req)
-				break
-			}
+// pull (up to ClusterChBufferSize) node lists off the fetchedNodesCh
+// keep only the latest list of nodes from the fetcher
+func (c *Cluster) getLatestNodesList() []Node {
+	var currentNodes []Node
+	haveFetchedNodes := false
+	i := 0
+LOOP:
+	for ; i < c.bufferSize; i++ {
+		select {
+		case currentNodes = <-c.fetchedNodesCh:
+			haveFetchedNodes = true
+		default:
+			break LOOP
 		}
 	}
+
+	if !haveFetchedNodes {
+		return c.getNodes()
+	}
+	return currentNodes
 }
 
-func (c *Cluster) closeSubscription(s *subscriber) {
-	c.reqCh <- s.inCh
+// addToCurrentNodeUpdates put the node updates on the nodes updates channel.
+func (c *Cluster) addToCurrentNodeUpdates(updates []NodeUpdate) {
+	c.NodesUpdatesCh <- updates
+	if len(updates) > 0 {
+		log.Infof("cluster has %d new node updates", len(updates))
+		// record time since last time saw node
+		c.stat.Gauge(stats.ClusterNodeUpdateFreqMs).Update(time.Since(c.priorNodeUpdateTime).Milliseconds())
+		c.priorNodeUpdateTime = time.Now()
+	}
 }
 
-func (c *Cluster) current() []Node {
-	var r []Node
-	for _, v := range c.state.nodes {
-		r = append(r, v)
+// GetNodes get (a copy of) the list of nodes last fetched by fetcher
+func (c *Cluster) getNodes() []Node {
+	ret := []Node{}
+	for _, node := range c.state.nodes {
```

Comment:
using built-int copy and/or preallocating your slice will be a big performance boost. Otherwise go will create many new new arrays behind the scenes( one on each append).